### PR TITLE
Fix some MoE routers

### PR DESCRIPTION
**Pull request by IlyasMoutawwakil**
# What does this PR do?

Same as https://github.com/huggingface/transformers/pull/43288
Also fixes phimoe and its integration tests

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/hunyuan_v1_moe/modular_hunyuan_v1_moe.py [hunk 1]
```diff
@@ -144,11 +144,6 @@ def route_tokens_to_experts(self, hidden_states):
         routing_weights = F.softmax(hidden_states, dim=1, dtype=torch.float)
         routing_weights, selected_experts = torch.topk(routing_weights, self.top_k, dim=-1)
         routing_weights /= routing_weights.sum(dim=-1, keepdim=True)
-        routing_weights = torch.zeros_like(hidden_states, dtype=torch.float32).scatter_(
-            1, selected_experts, routing_weights
-        )
-        return selected_experts, routing_weights.to(hidden_states.dtype)
-
         return selected_experts, routing_weights.to(hidden_states.dtype)
```
Comment (IlyasMoutawwakil) [inline src/transformers/models/hunyuan_v1_moe/modular_hunyuan_v1_moe.py:147]: it's interesting that this dead code is not caught by styling

File: src/transformers/models/phimoe/modeling_phimoe.py [hunk 2]
```diff
@@ -505,11 +506,8 @@ def forward(self, hidden_states: torch.Tensor) -> tuple[torch.Tensor, torch.Tens
             )
         router_logits = super().forward(hidden_states)
         routing_weights, selected_experts = sparsemixer(
-            router_logits,
-            jitter_eps=self.router_jitter_noise,
-            training=self.training,
+            router_logits, jitter_eps=self.router_jitter_noise, training=self.training, top_k=self.top_k
```
Comment (IlyasMoutawwakil) [inline src/transformers/models/phimoe/modeling_phimoe.py:504]: top_k was not passed and defaulted to 2

**Comment by IlyasMoutawwakil** [comment]
run-slow: hunyuan_v1_moe, phimoe

**Comment by IlyasMoutawwakil** [comment]
both models integration tests pass locally now

File: src/transformers/models/phimoe/modeling_phimoe.py [hunk 3]
```diff
@@ -578,8 +550,12 @@ def __init__(self, config: PhimoeConfig, layer_idx: int):
...
 
         self.mlp = PhimoeSparseMoeBlock(config)
-        self.input_layernorm = PhimoeRMSNorm(config.hidden_size, eps=config.rms_norm_eps)
-        self.post_attention_layernorm = PhimoeRMSNorm(config.hidden_size, eps=config.rms_norm_eps)
+
+        # Phimoe uses nn.LayerNorm
+        self.input_layernorm = nn.LayerNorm(config.hidden_size, eps=config.rms_norm_eps, elementwise_affine=True)
+        self.post_attention_layernorm = nn.LayerNorm(
+            config.hidden_size, eps=config.rms_norm_eps, elementwise_affine=True
+        )
```
Comment (IlyasMoutawwakil) [inline src/transformers/models/phimoe/modeling_phimoe.py:558]: Phimoe uses nn.LayerNorm with bias

File: src/transformers/integrations/moe.py [hunk 4]
```diff
@@ -105,27 +105,12 @@ def batched_mm_experts_forward(
     num_top_k = top_k_index.size(-1)
     num_tokens = hidden_states.size(0)
     hidden_dim = hidden_states.size(-1)
-    num_experts = self.gate_up_proj.size(0)
-
-    # Flatten top_k_index to get expert_ids per selected sample
-    expert_ids = top_k_index.reshape(-1)
-    token_idx = torch.arange(num_tokens, device=device).unsqueeze(1).expand(-1, num_top_k).reshape(-1)
-
-    # Resolve routing weights per selected sample, allowing top_k_weights to be either:
```
Comment (vasqu) [inline src/transformers/integrations/moe.py:114]: Glad to remove this :pray:

File: src/transformers/models/phimoe/modular_phimoe.py [hunk 5]
```diff
@@ -332,7 +330,14 @@ def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
...
 class PhimoeDecoderLayer(MixtralDecoderLayer):
-    pass
+    def __init__(self, config: PhimoeConfig, layer_idx: int):
+        super().__init__(config, layer_idx)
+
+        # Phimoe uses nn.LayerNorm
+        self.input_layernorm = nn.LayerNorm(config.hidden_size, eps=config.rms_norm_eps, elementwise_affine=True)
+        self.post_attention_layernorm = nn.LayerNorm(
+            config.hidden_size, eps=config.rms_norm_eps, elementwise_affine=True
+        )
```
Comment (vasqu) [inline src/transformers/models/phimoe/modular_phimoe.py:340]: Oh wow, that's an insane find - this model must have been broken for a long time

File: tests/models/hunyuan_v1_moe/test_modeling_hunyuan_v1_moe.py [hunk 6]
```diff
@@ -88,12 +84,10 @@ def tearDown(self):
 
     @slow
     def test_model_generation(self):
-        # we will compele this when model file change over
-        # pass
```
Comment (vasqu) [inline tests/models/hunyuan_v1_moe/test_modeling_hunyuan_v1_moe.py:92]: I have a PR here #43411 which fixes some wrong RoPE init, will this still work with that fix?

File: tests/models/phimoe/test_modeling_phimoe.py [hunk 7]
```diff
@@ -144,8 +144,8 @@ def test_model_phimoe_instruct_logits(self):
 
         EXPECTED_OUTPUT = torch.tensor(
             [
-                    [-3.4844, -2.4531, -1.1719, 0.6055, -0.4922, -0.1001, 0.8086, -0.2422, 0.3477, -1.0078],
-                    [-0.9766, 0.1631, -0.5508, 2.3594, 0.7031, 3.1719, 0.4141, 0.2305, 0.6055, -2.1250],
+                [-3.4844, -2.4688, -1.1719, 0.5703, -0.4902, -0.0942, 0.7773, -0.2539, 0.3223, -1.0234],
+                [-0.9805, 0.0811, -0.5273, 2.3438, 0.6914, 3.0781, 0.3164, 0.2197, 0.5312, -2.1094],
```
Comment (vasqu) [inline tests/models/phimoe/test_modeling_phimoe.py]: I guess that's the gpu diff between t4 and a10

**Review by vasqu — APPROVED** [review]
Thanks for checking another set of models, I guess we have them all fixed with this? 

The phimoe one is an interesting case :sweat_smile:

File: tests/models/phimoe/test_modeling_phimoe.py (hunk 7 above)
Comment (IlyasMoutawwakil) [reply tests/models/phimoe/test_modeling_phimoe.py]: ah yes, rerunning with eager experts impl to see if grouped is also contributing to the diff.

Comment (IlyasMoutawwakil) [reply tests/models/phimoe/test_modeling_phimoe.py]: hmm, grouped and eager are equivalent in terms of logits here (on A100). should I revert this change ? or maybe use the expectation class.

Comment (vasqu) [reply tests/models/phimoe/test_modeling_phimoe.py]: We should check against A10 runners - let's revert for now and check with run-slow first. If it needs a change, I can update it

File: tests/models/hunyuan_v1_moe/test_modeling_hunyuan_v1_moe.py (hunk 6 above)
Comment (IlyasMoutawwakil) [reply tests/models/hunyuan_v1_moe/test_modeling_hunyuan_v1_moe.py:92]: i guess, here I only removed a comment, I might have misunderstood but the initialization added in the PR is already in the __init__ of the class, why is it necessary in __init_weights__ as well ?

File: src/transformers/integrations/moe.py (hunk 4 above)
Comment (IlyasMoutawwakil) [reply src/transformers/integrations/moe.py:114]: yeah much cleaner !

File: src/transformers/models/phimoe/modular_phimoe.py (hunk 5 above)
Comment (IlyasMoutawwakil) [reply src/transformers/models/phimoe/modular_phimoe.py:340]: yeh ppl must've continued to use the remote code one

**Comment by IlyasMoutawwakil** [comment]
run-slow: hunyuan_v1_moe, phimoe

File: tests/models/hunyuan_v1_moe/test_modeling_hunyuan_v1_moe.py (hunk 6 above)
Comment (vasqu) [reply tests/models/hunyuan_v1_moe/test_modeling_hunyuan_v1_moe.py:92]: Apparently, it doesn't matter anymore what you init in `__init__` and `_init_weights` will overwrite in any case - meaning that if there is custom logic in init, it will not be applied. I want to refactor this so that it is no longer the case or rather that we depend on another init function for rope that allows users to do whatever they want to as init

**Comment by IlyasMoutawwakil** [comment]
the ci is struggling with model loading again 😭, even though `device_map="auto"`

File: src/transformers/integrations/moe.py [hunk 8]
```diff
@@ -105,27 +105,12 @@ def batched_mm_experts_forward(
...
-        sample_weights = top_k_weights[token_idx, expert_ids]
-    else:
-        raise ValueError(
-            f"top_k_weights has an invalid/unsupported shape. It should be either (num_tokens, num_top_k)({num_tokens}, {num_top_k}) "
-            f"or (num_tokens, num_experts)({num_tokens}, {num_experts}), but got {top_k_weights.shape}."
-        )
-    sample_weights = sample_weights.reshape(-1, 1)  # (S, 1)
+
+    # Reshape for easier indexing
+    # S is the number of selected tokens-experts pairs (S = num_tokens * num_top_k)
```
Comment (IlyasMoutawwakil) [inline src/transformers/integrations/moe.py:110]: clarification taken from #43439
